python/lsst/pipe/tasks/hips.py

```diff
@@ -604,3 +614,907 @@ def build_quantum_graph(
+            Image to write.
+        png_mapping : `astropy.visualization.lupton_rgb.AsinhMapping`
+            Mapping to convert image to scaled png.
+        shift_order : `int`, optional
+            HPX shift_order.
+        """
+        # WARNING: In general PipelineTasks are not allowed to do any outputs
+        # outside of the butler.  This task has been given (temporary)
+        # Special Dispensation because of the nature of HiPS outputs until
+        # a more controlled solution can be found.
+
+        dir_number = (pixel//10000)*10000
+        hips_dir = hips_base_path.join(
+            f'Norder{order}',
+            forceDirectory=True
+        ).join(
+            f'Dir{dir_number}',
+            forceDirectory=True
+        )
+
+        wcs = makeHpxWcs(order, pixel, shift_order=shift_order)
+
+        uri = hips_dir.join(f'Npix{pixel}.fits')
+
+        with ResourcePath.temporary_uri(suffix=uri.getExtension()) as temporary_uri:
+            image.writeFits(temporary_uri.ospath, metadata=wcs.getFitsMetadata())
+
+            uri.transfer_from(temporary_uri, transfer="copy", overwrite=True)
+
+        # And make a grayscale png as well
+
+        vals = 255 - png_mapping.map_intensity_to_uint8(image.array).astype(np.uint8)
+        vals[~np.isfinite(image.array) | (image.array < 0)] = 0
+        im = Image.fromarray(vals[::-1, :], "L")
+
+        uri = hips_dir.join(f'Npix{pixel}.png')
+
+        with ResourcePath.temporary_uri(suffix=uri.getExtension()) as temporary_uri:
+            im.save(temporary_uri.ospath)
+
+            uri.transfer_from(temporary_uri, transfer="copy", overwrite=True)
+
+    def _write_hips_color_png(
+            self,
+            hips_base_path,
+            order,
+            pixel,
+            image_red,
+            image_green,
+            image_blue,
+            png_mapping,
+    ):
+        """Write a color png HiPS image.
+
+        Parameters
+        ----------
+        hips_base_path : `lsst.resources.ResourcePath`
+            Resource path to the base of the HiPS directory tree.
+        order : `int`
+            HEALPix order of the HiPS image to write.
+        pixel : `int`
+            HEALPix pixel of the HiPS image.
+        image_red : `lsst.afw.image.Image`
+            Input for red channel of output png.
+        image_green : `lsst.afw.image.Image`
+            Input for green channel of output png.
+        image_blue : `lsst.afw.image.Image`
+            Input for blue channel of output png.
+        png_mapping : `astropy.visualization.lupton_rgb.AsinhMapping`
+            Mapping to convert image to scaled png.
+        """
+        # WARNING: In general PipelineTasks are not allowed to do any outputs
+        # outside of the butler.  This task has been given (temporary)
+        # Special Dispensation because of the nature of HiPS outputs until
+        # a more controlled solution can be found.
+
+        dir_number = (pixel//10000)*10000
```

Comment:
Consider factoring out this hips_dir, dir_number calculation since it is reused and is required to match.